################################################################################
#### This script uses GAPIT to calculate kinship matrices for use in GWAS. It is
#### configurable; kinship may be calculated on a genome-wide or per-chromosome
#### basis (the latter is used in the K-chromosome GWAS approach). Additionally,
#### the kinship algorithm to use (VanRaden or Zhang) may be specified.
####
#### The set of genotypes to use are the 46K SNPs that are pre-pruned by LD.
####
#### Last updated 12/22/2016
################################################################################

################################################################################
#### WORKSTATION CONFIGURATION
################################################################################

#### Laptop
workstation.topLevel <- "/home/james/GoreLab"

#### Server
#workstation.topLevel <- "/home/jamesc"

################################################################################
#### DEPENDENCIES
####
#### This script also requires TASSEL5 be installed; the command line pipeline
#### is used to perform intermediate HapMap functions. The default directory in
#### which it looks for the TASSEL5 installation is ~/TASSEL5/
####
#### For the chromosome-wise calculation, this script also calls a python script
#### called 'PatchTaxaNames.py'. This script corrects modifications to the
#### genotype names that result from reading header=TRUE for this configuration.
################################################################################

## at least one of these is required to source the local copy of gapit functions
## not sure which. local copy of gapit functions is required for this script.
## again I don't know why. 
library(multtest)
library(gplots)
library(LDheatmap)
library(genetics)
library(compiler)
library(scatterplot3d)
library(EMMREML)

#source("http://zzlab.net/GAPIT/gapit_functions.txt")
source(paste(workstation.topLevel,"/GAPIT/gapit_functions.txt",sep=""))
source("http://www.zzlab.net/GAPIT/emma.txt")

patchScript.filename <- paste(workstation.topLevel,"/MaizeLeafCuticle/GWAS/pipeline/PatchTaxaNames.py",sep="")

################################################################################
#### CONFIGURATION
####
#### This script is tailored for the maize leaf cuticle project. While the base
#### set of genotypes to use for kinship is common among all iterations (SNPs
#### derived from Hirsch et al., 2014), the particular taxa over which to
#### calculate kinship depends on the phenotype of interest, because the taxa
#### for which scores are available vary by phenotype. The minimum parameters to
#### determine the base genotype FILE (already filtered for the appropriate
#### taxa) are: whether the phenotype of interest is absolute cuticular
#### evaporation or relative cuticular evaporation (rate scaled by leaf dry
#### mass); and the environments from which to include data. These base genotype
#### files should already have been created prior to calling this script. They
#### are generated by the script "BuildHapMapsByPhenotype.py". Note that this
#### script is agnostic to the particular BLUX model used to generate GWAS
#### phenotypes, because the taxa included are invariant prior to model
#### selection.
################################################################################

#### Project-specific GWAS top-level
parent.dir <- paste(workstation.topLevel,"/MaizeLeafCuticle/GWAS/",sep="")

#### Parameters specifying the phenotype
scaleByLeafSize <- TRUE
#envs <- "AZ16"
#envs <- "SD16"
envs <- "AllEnvs"

#### Algorithm used to calculate kinship matrices
kin.alg <- "VanRaden"
#kin.alg <- "Zhang"

#### For K-chromosome
chr.count <- 10

################################################################################
#### Backend functions
################################################################################

#### Function abstracting common setup routines: determining the name of the
#### target directory, input genotype filename, and output filename for the
#### kinship matrix.
setupVariables <- function(parent.dir, scaleByLeafSize, envs) {
  target.dir <- parent.dir
  if (scaleByLeafSize) {
    target.dir <- paste(target.dir,"RelativeCuticularEvaporation",sep="")
  } else {
    target.dir <- paste(target.dir,"AbsoluteCuticularEvaporation",sep="")
  }
  genotype.filename <- paste(target.dir,"/",envs,"/MLC_taxa_imputed_46K_filtered_LDPruned_phenoSpecific.hmp.txt",sep="")
  target.dir <- paste(target.dir,"/",envs,sep="")
  returnVar <- list(target.dir,genotype.filename)
  return(returnVar)
}

#### Function to calculate genome-wise kinship matrices.
buildGenomeWiseKinshipMatrices <- function(parent.dir, scaleByLeafSize, envs, kin.alg) {
  
  ## Setup subroutine
  setupVars = setupVariables(parent.dir, scaleByLeafSize, envs)
  target.dir = setupVars[[1]]
  genotype.filename = setupVars[[2]]
  
  ## Choose a descriptive folder name into which to kinship matrix files, then
  ## create and switch into that folder.
  setwd(paste(target.dir,"KinshipMatrices",sep="/"))
  export.dir = paste("GenomeWide_",kin.alg,sep="")
  try(system(paste("mkdir","-p",export.dir,sep=' ')))
  setwd(paste(target.dir,"KinshipMatrices",export.dir,sep="/"))
  
  ## Call GAPIT to create the kinship matrices
  genotypes.for.kinship = read.table(genotype.filename, sep="\t", header = FALSE)
  res = GAPIT( G = genotypes.for.kinship, kinship.cluster = c("average"), kinship.group = c("Mean"), kinship.algorithm = kin.alg)
}

#### Function to calculate chromosome-wise kinship matrices
buildChromosomeWiseKinshipMatrices <- function(parent.dir, scaleByLeafSize, envs, kin.alg, chr.count) {
  
  ## Setup subroutine
  setupVars = setupVariables(parent.dir, scaleByLeafSize, envs)
  target.dir = setupVars[[1]]
  genotype.filename = setupVars[[2]]
  
  ## Choose a descriptive folder name into which to kinship matrix files, then
  ## create and switch into that folder.
  setwd(paste(target.dir,"KinshipMatrices",sep="/"))
  export.dir = paste("K_Chromosome_",kin.alg,sep="")
  try(system(paste("mkdir","-p",export.dir,sep=' ')))
  setwd(paste(target.dir,"KinshipMatrices",export.dir,sep="/"))
  
  ## Determine the number of taxa in the input genotype file
  header.line = readLines(genotype.filename, n=1)
  header.vals = strsplit(header.line,"\t")[[1]]
  taxa.count = length(header.vals) - 11
  
  ## Read in the genotype file
  geno = read.table(genotype.filename, sep="\t", header=TRUE, colClasses=c(rep('character',2),rep('numeric',2),rep('character',7+taxa.count)))
  
  for (j in c(1:chr.count)) {
  #for (j in c(1,2,3,4,5,6,7)) {
    
    genotmp = geno[-which(geno$chrom == j),]
    g3 = genotmp$chrom
    gg3 = g3[which(g3 > j)]
    gg3 = gg3 - 1
    g3 = c(g3[which(g3 < j)], gg3)
    genotmp[,3] = g3
    # reindexing factors because R won't automatically do this when taking table subset
    write.table(genotmp, file='genotmp', row.names=FALSE, col.names=TRUE, sep='\t', quote=FALSE)
    genotmp <- read.table('genotmp', header=FALSE, sep='\t')
    res = GAPIT( G = genotmp, kinship.cluster = c("average"), kinship.group = c("Mean"), kinship.algorithm = kin.alg)
    system("rm genotmp")
    
    ## Convert to tab-delimited file
    kin.filename = paste(target.dir,"/KinshipMatrices/",export.dir,"/GAPIT.Kin.",kin.alg,".csv",sep="")
    system(paste ("sed -i -e 's/,/\t/g' ", kin.filename, sep=""))
    
    ## Correct genotype name modifications resulting from R's automatic header field
    ## conversion, routing to a new file
    new.kin.filename = paste(target.dir,"/KinshipMatrices/",export.dir,"/",kin.alg,'_Kinship_chr', j, '.txt', sep='')
    system(paste ("python",patchScript.filename,kin.filename,new.kin.filename,sep=" "))
  }

}
################################################################################
#### Run the script
################################################################################
buildGenomeWiseKinshipMatrices(parent.dir, scaleByLeafSize, envs, kin.alg)

#buildChromosomeWiseKinshipMatrices(parent.dir, scaleByLeafSize, envs, kin.alg, chr.count)

